service/mcs/persistent/src/main/java/eu/europeana/cloud/service/mcs/persistent/cassandra/CassandraDataSetDAO.java

```diff
@@ -681,46 +685,83 @@ private Representation mapRowToRepresentationStub(Row row){
                 representationName, cloudId);
         ResultSet rs = connectionProvider.getSession().execute(boundStatement);
         QueryTracer.logConsistencyLevel(boundStatement, rs);
     }
 
-    public List<String> getDataSetsRevisionWithPagination(String providerId, String datasetId, String revisionId, String representationName, String startFromCloudId, int limit){
-        String providerDatasetId = createProviderDataSetId(providerId,datasetId);
-        BoundStatement boundStatement = getDataSetsRevisionWithPagination.bind(providerDatasetId,revisionId,representationName,startFromCloudId,limit);
+//    public List<String> getDataSetsRevisionWithPagination(String providerId, String datasetId, String revisionId, String representationName, String startFromCloudId, int limit){
+//        String providerDatasetId = createProviderDataSetId(providerId,datasetId);
+//        BoundStatement boundStatement = getDataSetsRevisionWithPagination.bind(providerDatasetId,revisionId,representationName,startFromCloudId,limit);
+//        ResultSet rs = connectionProvider.getSession().execute(boundStatement);
+//        QueryTracer.logConsistencyLevel(boundStatement, rs);
+//        List<String> cloudIds = new ArrayList<>();
+//        for (Row r : rs){
+//            String cloudId = r.getString("cloud_Id");
+//            cloudIds.add(cloudId);
+//        }
+//        return cloudIds;
+//    }
+
+//    public List<String> getDataSetsRevision(String providerId, String datasetId, String revisionId, String representationName, int limit){
+//        String providerDatasetId = createProviderDataSetId(providerId,datasetId);
+//        BoundStatement boundStatement = getDataSetsRevision.bind(providerDatasetId,revisionId,representationName, limit);
+//        ResultSet rs = connectionProvider.getSession().execute(boundStatement);
+//        QueryTracer.logConsistencyLevel(boundStatement, rs);
+//        List<String> cloudIds = new ArrayList<>();
+//        for (Row r : rs){
+//            String cloudId = r.getString("cloud_Id");
+//            cloudIds.add(cloudId);
+//        }
+//        return cloudIds;
+//    }
+
+    public List<Properties> getDataSetsRevisions(String providerId, String dataSetId, String revisionProviderId, String revisionName, Date revisionTimestamp, String representationName, String nextToken, int limit) {
+        List<Properties> result = new ArrayList<>(limit);
+
+        // bind parameters, set limit to max int value
+        BoundStatement boundStatement = getDataSetsRevision.bind(providerId, dataSetId, revisionProviderId, revisionName, revisionTimestamp, representationName, Integer.MAX_VALUE);
+        // limit page to "limit" number of results
+        boundStatement.setFetchSize(limit);
+        // when this is not a first page call set paging state in the statement
+        if (nextToken != null)
+            boundStatement.setPagingState(PagingState.fromString(nextToken));
```

Comment:
I would also get rid of PaginingState and change the query to handle "cloudId >nextToken" ; since cloudId is also in the clustering keys .